### PR TITLE
Add a delete confirmation dialog for widgets

### PR DESCRIPTION
**Pull request by rxl881**
See - https://github.com/vector-im/riot-web/issues/5115

### CODE REVIEW

**Comment by lukebarnard1** [comment]
Needs linting!

File: src/components/views/elements/AppTile.js [hunk 1]
```diff
@@ -172,18 +172,32 @@ export default React.createClass({
...
-            ).then(() => {
-                console.log('Deleted widget');
-            }, (e) => {
-                console.error('Failed to delete widget', e);
-                this.setState({deleting: false});
+            // Show delete confirmation dialog
+            const QuestionDialog = sdk.getComponent("dialogs.QuestionDialog");
+            Modal.createTrackedDialog('Delete Widget', '', QuestionDialog, {
+                title: _t("Delete Widget"),
+                description:
```
Comment (lukebarnard1) [inline src/components/views/elements/AppTile.js]: I think the description can be a string

File: src/components/views/elements/AppTile.js [hunk 2]
```diff
@@ -172,18 +172,32 @@ export default React.createClass({
...
+                    <div>
+                 { _t("Deleting a widget removes it for all users in this room - are you sure you want to delete this widget?") }
+                    </div>,
+                button: _t("OK"),
+                onFinished: (confirmed) => {
+                    if (confirmed) {
```
Comment (lukebarnard1) [inline src/components/views/elements/AppTile.js]: Best to `if (!confirmed) return; `

File: src/components/views/elements/AppTile.js [hunk 3]
```diff
@@ -172,18 +172,32 @@ export default React.createClass({
...
+                        console.log("Delete widget %s", this.props.id);
+                        this.setState({deleting: true});
+                        MatrixClientPeg.get().sendStateEvent(
+                            this.props.room.roomId,
+                            'im.vector.modular.widgets',
+                            {}, // empty content
+                            this.props.id,
+                        ).then(() => {
+                            console.log('Deleted widget');
```
Comment (lukebarnard1) [inline src/components/views/elements/AppTile.js]: This log isn't very useful afaics, best to just `catch`?

File: src/components/views/elements/AppTile.js [hunk 4]
```diff
@@ -172,18 +172,32 @@ export default React.createClass({
...
-            }, (e) => {
-                console.error('Failed to delete widget', e);
-                this.setState({deleting: false});
+            // Show delete confirmation dialog
+            const QuestionDialog = sdk.getComponent("dialogs.QuestionDialog");
+            Modal.createTrackedDialog('Delete Widget', '', QuestionDialog, {
+                title: _t("Delete Widget"),
+                description:
+                    <div>
+                 { _t("Deleting a widget removes it for all users in this room - are you sure you want to delete this widget?") }
```
Comment (lukebarnard1) [inline src/components/views/elements/AppTile.js]: Could split this over multiple lines, but I guess the linter caught this

File: src/components/views/elements/AppTile.js (hunk 1 above)
Comment (rxl881) [reply src/components/views/elements/AppTile.js]: I'm confused by this comment. I'm not sure what change you are suggesting.

This is very similar to usage in places like - https://github.com/matrix-org/matrix-react-sdk/blob/master/src/components/structures/UserSettings.js#L427-L433

Comment (t3chguy) [reply src/components/views/elements/AppTile.js]: You could just pass the output of _t rather than wrap it in a div

Comment (rxl881) [reply src/components/views/elements/AppTile.js]: I see. Thanks for the clarification @t3chguy. I've removed the div, as suggested ... but it seems a bit inconsistent with the usage in lots of the rest of the code ...

```
components/views/login/RegistrationForm.js-                    description:
components/views/login/RegistrationForm.js-                        <div>
--
components/views/messages/TextualBody.js-                description:
components/views/messages/TextualBody.js-                    <div>
--
components/views/rooms/MemberInfo.js-                    description:
components/views/rooms/MemberInfo.js-                        <div>
--
components/views/rooms/MessageComposer.js-            description: (
components/views/rooms/MessageComposer.js-                <div>
--
components/views/rooms/RoomSettings.js-            description:
components/views/rooms/RoomSettings.js-                <div>
--
components/views/rooms/RoomSettings.js-            description: (
components/views/rooms/RoomSettings.js-                <div>
--
components/views/settings/ChangePassword.js-            description:
components/views/settings/ChangePassword.js-                <div>
...
```

**Comment by rxl881** [comment]
> Needs linting!

Out of interest, were you just referring to the one long string? That's the only thing that was caught by the linter (or that I can see from your comments).

The long string was just shown as a warning locally ("max-len": ["warn", in eslintrc), and I assumed that long string literals were probably OK to leave as warnings? but this seems to have caused the travis build to fail, which is a little odd (it would be good to know where that config is coming from) ...

Assuming that string literals do always need to be broken up, is our preferred approach just to break it in to relatively arbitrary fragments and concatenate, like I've done here? Cheers.

**Comment by lukebarnard1** [comment]
It looks like this was the only thing caught by the linter, yep.

We generally split strings as such:
```js
_t(
    "Some really long string " +
    "some other long string."
)
```

Lint warnings will cause the build to fail, although it is confusing that both warnings and errors will result in the build failing. The ESlint config is `.eslintrc.js`.

**Comment by rxl881** [comment]
Thanks @lukebarnard1.

I was just wondering where the config lived that is causing travis-ci build to fail on warnings, but I guess that doesn't really matter now that I know that it does (fail on warnings) :)

Is this now good to go? Cheers.

File: src/components/views/elements/AppTile.js [hunk 5]
```diff
@@ -172,18 +172,28 @@ export default React.createClass({
...
-                console.log('Deleted widget');
-            }, (e) => {
-                console.error('Failed to delete widget', e);
-                this.setState({deleting: false});
+            // Show delete confirmation dialog
+            const QuestionDialog = sdk.getComponent("dialogs.QuestionDialog");
+            Modal.createTrackedDialog('Delete Widget', '', QuestionDialog, {
+                title: _t("Delete Widget"),
+                description: _t("Deleting a widget removes it for all users in this room." +
+                " Are you sure you want to delete this widget?"),
```
Comment (lukebarnard1) [inline src/components/views/elements/AppTile.js]: I'd generally indent like this:
```js
_t(
    "Some really long string " +
    "some other long string."
)
```

**Comment by lukebarnard1** [comment]
LGTM apart from indentation on that split string

File: src/components/views/elements/AppTile.js (hunk 5 above)
Comment (rxl881) [reply src/components/views/elements/AppTile.js]: Thanks @lukebarnard1 - done 👍

**Pull request merged**